scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt

```diff
@@ -0,0 +1,209 @@
+ * If information about a Git repository in GitHub is available, this is used. Otherwise, the coordinates
+ * are derived from the identifier.
+ */
+private fun packageCoordinates(
+    id: Identifier,
+    vcs: VcsInfo?,
+    sourceArtifact: RemoteArtifact?
+): ClearlyDefinedService.Coordinates {
+    val sourceLocation = id.toClearlyDefinedSourceLocation(vcs?.toVcsInfoCurationData(), sourceArtifact)
+    return sourceLocation?.toCoordinates() ?: id.toClearlyDefinedCoordinates()
+}
+
+/**
+ * Search in the given list of [tools] for the entry for ScanCode that has processed the
+ * [packageUrl]. If such an entry is found, the version of ScanCode is returned; otherwise, result is null.
+ */
+private fun findScanCodeVersion(tools: List<String>, packageUrl: String): String? {
+    val toolUrl = "$packageUrl/$TOOL_SCAN_CODE/"
+    return tools.find { it.startsWith(toolUrl) }?.substring(toolUrl.length)
+}
+
+/**
+ * Return an empty success result for the given [id].
+ */
+private fun emptyResult(id: Identifier): Result<ScanResultContainer> =
+    Success(ScanResultContainer(id, listOf()))
+
+/**
+ * A storage implementation that tries to download ScanCode results from ClearlyDefined.
+ *
+ * This storage implementation maps the requested package to coordinates used by ClearlyDefined. It then tries
+ * to find a harvested ScanCode result for these coordinates using the ClearlyDefined REST API.
+ */
+class ClearlyDefinedStorage(
+    /** The configuration for this storage implementation. */
+    val configuration: ClearlyDefinedStorageConfiguration
+) : ScanResultsStorage() {
+    /** The service for interacting with ClearlyDefined. */
+    private val clearlyDefinedService = ClearlyDefinedService.create(configuration.serverUrl)
+
+    override fun readFromStorage(id: Identifier): Result<ScanResultContainer> =
+        readPackageFromClearlyDefined(id, null, null)
+
+    override fun readFromStorage(pkg: Package, scannerDetails: ScannerDetails): Result<ScanResultContainer> =
+        readPackageFromClearlyDefined(pkg.id, pkg.vcs, pkg.sourceArtifact.takeIf { it.url.isNotEmpty() })
+
+    override fun addToStorage(id: Identifier, scanResult: ScanResult): Result<Unit> =
+        Failure("Storing results in ClearlyDefined is not supported.")
+
+    /**
+     * Try to obtain a [ScanResult] produced by ScanCode from ClearlyDefined for the package with the given [id].
+     * Use the VCS information [vcs] and an optional [sourceArtifact] to determine the coordinates for looking up
+     * the result.
+     */
+    private fun readPackageFromClearlyDefined(
+        id: Identifier,
+        vcs: VcsInfo?,
+        sourceArtifact: RemoteArtifact?
+    ): Result<ScanResultContainer> =
+        try {
+            readFromClearlyDefined(id, packageCoordinates(id, vcs, sourceArtifact))
+        } catch (e: IllegalArgumentException) {
+            log.warn("Could not obtain ClearlyDefined coordinates for package $id", e)
```

Comment:
Please use `'${id.toCoordinates()}'` when printing `Identifier`s. Also: Missing dot at end of sentence 😀